pkg/devfile/adapters/common/utils.go

```diff
@@ -20,3 +31,160 @@ func GetSupportedComponents(data versions.DevfileData) []common.DevfileComponent
+		command := GetCommand(data, devfileRunCmd)
+		if !reflect.DeepEqual(emptyCommand, command) {
+			for _, action := range command.Actions {
+				components = append(components, *action.Component)
+			}
+		}
+	} else {
+		for _, command := range GetSupportedCommands(data) {
+			if strings.Contains(command.Name, DefaultDevfileRunCommand) {
+				for _, action := range command.Actions {
+					components = append(components, *action.Component)
+				}
+			}
+		}
+	}
+
+	return components
+}
+
+// ValidateAndGetPushDevfileCommands returns a build and run command. It checks if a build
+// or a run command has been explicitly provided during odo push, otherwise it
+// iterates through the devfile commands and returns the default supported command. If neither
+// a build command nor a run command is found, it throws an error
+func ValidateAndGetPushDevfileCommands(data versions.DevfileData, devfileBuildCmd, devfileRunCmd string) ([]common.DevfileCommand, error) {
+	var pushDevfileCommands []common.DevfileCommand
+	var emptyCommand common.DevfileCommand
+	validateBuildCommand := false
+	validateRunCommand := false
+
+	devfileSupportedCommands := GetSupportedCommands(data)
+
+	// Validate the build command if it was provided during odo push
+	if devfileBuildCmd != "" {
+		devfileCommand := GetCommand(data, devfileBuildCmd)
+		if !reflect.DeepEqual(emptyCommand, devfileCommand) {
+			pushDevfileCommands = append(pushDevfileCommands, devfileCommand)
+			validateBuildCommand = true
+		}
+	}
+
+	// Validate the run command if it was provided during odo push
+	if devfileRunCmd != "" {
+		devfileCommand := GetCommand(data, devfileRunCmd)
+		if !reflect.DeepEqual(emptyCommand, devfileCommand) {
+			pushDevfileCommands = append(pushDevfileCommands, devfileCommand)
+			validateRunCommand = true
+		}
+	}
+
+	// If neither is validated, iterate the devfile commands to see for a list of supported commands
+	for _, supportedCommand := range devfileSupportedCommands {
+		if reflect.DeepEqual(supportedCommand.Name, DefaultDevfileBuildCommand) && !validateBuildCommand {
+			pushDevfileCommands = append(pushDevfileCommands, supportedCommand)
+			validateBuildCommand = true
+		}
+		if reflect.DeepEqual(supportedCommand.Name, DefaultDevfileRunCommand) && !validateRunCommand {
+			pushDevfileCommands = append(pushDevfileCommands, supportedCommand)
+			validateRunCommand = true
+		}
+	}
+
+	if !validateBuildCommand || !validateRunCommand {
+		return []common.DevfileCommand{}, fmt.Errorf("devfile command validation failed, validateBuildCommand: %v validateRunCommand: %v", validateBuildCommand, validateRunCommand)
+	}
+
+	return pushDevfileCommands, nil
+}
+
+// IsDevfileCommandSupported checks if a devfile command is supported by default
+func IsDevfileCommandSupported(commandName string) bool {
+	isSupported := false
+
+	switch commandName {
+	case DefaultDevfileBuildCommand:
+		fallthrough
+	case DefaultDevfileRunCommand:
+		fallthrough
+	case DefaultDevfileDebugCommand:
```

Comment:
I see we're adding in cases for `DefaultDevfileDebugCommand` and `DefaultDevfileTestCommand`. Should the Push function be updated to take those in too? Right now it only accepts `[]string, devfileBuildCmd, devfileRunCmd string

Reply:
We will be having a new command for test and debug. They will not be executed through odo push. I am removing the debug and test command name constants to avoid confusion. They can be re-added later on